### DISCUSSION TITLE
Why Doesn’t Babel Allow Specifying Detailed Versions for Core-js in Plugin Options?

### DISCUSSION THREAD

**Discussion by yceffort** [babel/babel#16944] [Q&A] [answered]
In the options for `plugin-transform-runtime`, the version I can specify for `core-js` is either 2, 3, or false.

https://github.com/babel/babel/blob/8fa44adcad55036805e3d9548893c138dc4dcd64/packages/babel-plugin-transform-runtime/src/babel-7/polyfills.cjs#L49-L55

And based on this value, it determines which plugin to use.

https://github.com/babel/babel/blob/8fa44adcad55036805e3d9548893c138dc4dcd64/packages/babel-plugin-transform-runtime/src/babel-7/polyfills.cjs#L81-L90

Before this code, you can see that the plugin options are declared.

https://github.com/babel/babel/blob/8fa44adcad55036805e3d9548893c138dc4dcd64/packages/babel-plugin-transform-runtime/src/babel-7/polyfills.cjs#L69-L79

In these options, no specific version is specified. For example, with `core-js` version 3, it will use `babel-plugin-polyfill-corejs3`, and since the version is not explicitly stated in the documentation, the default value of 3.0 will be applied.

https://github.com/babel/babel-polyfills/tree/main/packages/babel-plugin-polyfill-corejs3

And I found that this version affects the list of polyfills that `babel-plugin-polyfill-corejs3` can provide.

https://github.com/babel/babel-polyfills/blob/bacb1dfd82abab2b484e98fb25a69798d8471f31/packages/babel-plugin-polyfill-corejs3/src/index.ts#L83

https://github.com/zloirock/core-js/blob/master/packages/core-js-compat/get-modules-list-for-target-version.js

Here’s my question: why are we restricted from specifying versions like the current latest, 3.38, instead of just "3"? If we could specify a more precise version, we could access a broader range of polyfills.

https://github.com/zloirock/core-js/blob/master/packages/core-js-compat/src/modules-by-versions.mjs

Apologies for the novice question about Babel. Any responses would be greatly appreciated and very helpful. Thank you!

**Comment by liuxingbaoyu** [comment]
Thanks for your report, this is an issue with our implementation.
As a workaround, you can use `babel-plugin-polyfill-corejs3` directly.

**Reply by yceffort** [reply]
Thank you for the quick response. Would you be open to allowing the option to pass versions directly, beyond just 2, 3, or `false`, in the future? If so, I'd be interested in making the code modifications. If not, I'll proceed with directly modifying `babel-plugin-polyfill-corejs3` as you suggested.

**Reply by liuxingbaoyu** [reply]
Thank you so much for your willingness to contribute.
This is a section I'm not familiar with, let's hear what @nicolo-ribaudo thinks. :)

**Reply by nicolo-ribaudo** [reply]
To be honest I would prefer to push folks directly to use `babel-plugin-polyfill-corejs3`, and eventually completely remove `core-js` support from `@babel/plugin-transform-runtime` (making it something just for helpers). Many people use it without actually needing the core-js integration.

**Reply by nicolo-ribaudo** [reply, marked as answer]
Oh actually, the Babel 8 alpha already removed corejs from `@babel/plugin-transform-runtime`, so using `babel-plugin-polyfill-corejs3` directly is definitely the recommended way!

**Reply by yceffort** [reply]
Thank you for your response.

```js
plugins: [
    ['@babel/plugin-transform-runtime'],
    [
        'babel-plugin-polyfill-corejs3',
        {
            method: 'usage-pure',
            version: '3.38.1',
            proposals: true,
        },
    ],
],
```

Following this approach, `@babel/plugin-transform-runtime` is used specifically to add Babel's helper functions, while `babel-plugin-polyfill-corejs3` injects the necessary polyfills at runtime. Please let me know if I’m misunderstanding this setup. This configuration is working well and matches my requirements.

---

As you pointed out, it seems this feature will be removed in Babel 8. While the release date for Babel 8 is unclear, if it’s coming soon, it seems best not to include this feature.

https://github.com/babel/babel/blob/8fa44adcad55036805e3d9548893c138dc4dcd64/packages/babel-plugin-transform-runtime/src/index.ts#L8-L9

---

One issue I’m currently dealing with is outlined [here](https://github.com/babel/babel-polyfills/issues/222): the `findLast` method isn't being polyfilled correctly, so I submitted an issue.

After investigating, I found that the issue arises because `instance/find-last` is missing from the [babel-runtime-corejs3-paths.ts file](https://github.com/babel/babel-polyfills/blob/bacb1dfd82abab2b484e98fb25a69798d8471f31/packages/babel-plugin-polyfill-corejs3/src/babel-runtime-corejs3-paths.ts).

However, using `babel-plugin-polyfill-corejs3` allows the polyfill to be applied based on `core-js-pure`, which resolves the issue.

https://github.com/babel/babel-polyfills/blob/bacb1dfd82abab2b484e98fb25a69798d8471f31/packages/babel-plugin-polyfill-corejs3/src/index.ts#L142-L146

https://github.com/babel/babel-polyfills/blob/bacb1dfd82abab2b484e98fb25a69798d8471f31/packages/babel-plugin-polyfill-corejs3/src/index.ts#L133-L138

I’m uncertain if this approach will also be deprecated with Babel 8, but I’ve decided not to rely on it either way.

Thank you for all your efforts in supporting Babel and the JavaScript ecosystem—I greatly appreciate it!

**Reply by nicolo-ribaudo** [reply]
Your new config matches exactly what you should do :)
